### PR TITLE
Aligned TF Reshape layer behaviour

### LINKED ISSUE #19489: Reshape layer of keras (tensorflow) is interpreted incorrectly by readNetFromTensorflow() in dnn
##### System information (version)
- OpenCV => 4.5.1 (tested on 4.4.0 as well)
- Operating System / Platform => Ubuntu 20.04
- Compiler => Pre-built library (for OpenCV 4.5.1) and gcc (for OpenCV 4.4.0)

##### Detailed description
I tried to use a deep learning model trained on keras (tensorflow) on OpenCV via dnn. The output of dnn was very different from that of keras, so I started debugging. It looks like the outputs are different because the Reshape layer of keras is interpreted incorrectly in OpenCV dnn. If I remove the Reshape layer the model works just fine. I searched as much as I can; I found some similar issues but I think this exact bug is not mentioned before.

##### Steps to reproduce
The code below should reproduce the error. You can also use [this .zip file](https://github.com/opencv/opencv/files/5953629/reproduce_reshape_bug.zip), which also contains the test.jpg file that's used below. As you can see, if you set the `add_reshape_layer` variable to `False`, the model will not contain the reshape layer and will work just fine -- the output of keras will match that of OpenCV dnn. 

```
import tensorflow.keras as keras
import tensorflow as tf
from tensorflow.python.framework.convert_to_constants import convert_variables_to_constants_v2
import numpy as np
import cv2

# If you turn this off the code will work fine -- the outputs of keras and OpenCV dnn will match
add_reshape_layer = True

# This is the simplified model that I constructed for debugging
model_simple = keras.models.Sequential([keras.layers.Conv2D(64, kernel_size=7, activation='relu', strides=(2,2), padding="same"),
    keras.layers.BatchNormalization(),
    keras.layers.Conv2D(64, kernel_size=3, activation='relu', strides=(2,2), padding="same"),
    keras.layers.BatchNormalization(),
    keras.layers.Conv2D(32, kernel_size=3, activation='relu', strides=(2,2), padding="same", use_bias=False),
    keras.layers.BatchNormalization(),
    keras.layers.Conv2D(12, kernel_size=3, activation='relu', padding="same", use_bias=False),
    keras.layers.BatchNormalization(),
    keras.layers.Conv2D(2, kernel_size=3, activation='relu', padding="same", use_bias=False),
    keras.layers.BatchNormalization(),
    keras.layers.Flatten(),
    keras.layers.Dense(16*16*6, activation='relu', use_bias=False)
])

if add_reshape_layer:
    model_simple.add(keras.layers.Reshape((16,16,6), input_shape=(16*16*6, )))

# Building and freezing model
model_simple.build(input_shape=(None,256,256,3))                                 
model_simple.trainable = False

model_simple.save('SIMPLE_MODEL')
loaded = tf.saved_model.load('SIMPLE_MODEL')
infer = loaded.signatures['serving_default']
f = tf.function(infer).get_concrete_function(conv2d_input=tf.TensorSpec(shape=[1, 256, 256, 3], dtype=tf.float32))
f2 = convert_variables_to_constants_v2(f)

graph_def = f2.graph.as_graph_def()

# Export frozen graph
with tf.io.gfile.GFile('simple_model.pb', 'wb') as f:
   f.write(graph_def.SerializeToString())


# Try on test image
im = cv2.imread('test.jpg')/255.0
rim = tf.reshape(im, (-1, 256, 256, 3))

# y_keras: output for test image from keras interface
y_keras = model_simple.predict(rim)

model_dnn = cv2.dnn.readNetFromTensorflow('./simple_model.pb')
model_dnn.setInput(cv2.dnn.blobFromImage(np.float32(im)))
y_dnn = model_dnn.forward()

if add_reshape_layer:
    print(np.max(np.abs(y_keras[0,:,:,0]-y_dnn[0,0,:,:]))/np.max(np.abs(y_keras))) # prints large difference
else:
    print(np.max(np.abs(y_keras-y_dnn))/np.max(np.abs(y_keras))) # prints very small difference -- practically zero
```

### PR DESCRIPTION
**Pull request by LupusSanctus**
Merge with extra: opencv/opencv_extra#857

Fixes #19489: [tf.keras.layers.Reshape](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Reshape) layer interpreted incorrectly by readNetFromTensorflow() in dnn.

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2021.1.0:20.04
build_image:Custom Win=openvino-2021.1.0
build_image:Custom Mac=openvino-2021.1.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/tensorflow/tf_importer.cpp [hunk 1]
```diff
@@ -295,6 +295,17 @@ DataLayout getDataLayout(
     return it != data_layouts.end() ? it->second : DATA_LAYOUT_UNKNOWN;
 }
 
+static
+bool hasAllOnes(const Mat &inputs, int startPos, int endPos)
+{
+    for (int i = startPos; i < endPos; i++)
+    {
+        if (inputs.at<int>(0, i) != 1 || inputs.at<int>(0, i) != -1)
```
Comment (alalek) [inline modules/dnn/src/tensorflow/tf_importer.cpp:303]: > `inputs.at<int>(0, i)`

Consider using single arg version instead: `inputs.at<int>(i)`

---

It makes sense to add extra checks before using of low-level processing:
```
CV_CheckLE(inputs.dims, 2, "");
CV_CheckGE(startPos, 0, "");
CV_CheckLE(startPos, endPos, "");
CV_CheckLT((size_t)endPos, inputs.total(), "");
```

**Review by alalek — COMMENTED** [review]
Thank you!

Please rebase patch on 3.4 branch (+extra).

File: modules/dnn/src/tensorflow/tf_importer.cpp (hunk 1 above)
Comment (LupusSanctus) [reply modules/dnn/src/tensorflow/tf_importer.cpp]: @alalek, thank you for the review!

**Review by alalek — APPROVED** [review]
Looks good to me! Thank you :+1:
